### ISSUE TITLE
Phase a vcf with haplotagged reads

### ISSUE THREAD

**Issue by mrvollger** [whatshap/whatshap#520]
Hi @tobiasmarschall and team!

It's on my todo list to write a tool that takes a bam with phased reads and then phases a vcf with that preexisting information. This often comes up for us when we phase reads with k-mers or an assembly, and we want to create a phased vcf to match.

I was going to make a standalone tool, but then I thought that depending on how things are set up in Whatshap, much of the work might already be done.

If I could get some pointers on where to hook into the codebase, I'd try to make a PR. However, I have to admit that my C/C++ skills are awful, so I'd probably only be able to help if I could do it through the Python side of things.

Cheers,
Mitchell

**Comment by schrins** [comment]
Hello Mitchell,

I cannot speak for the entire team here, but I just want to drop some thoughts. The usecase definitely sounds interesting and some of the existing code could definitely save some work. The way the `phase` (and `polyphase`) commands work is, that they first read in a given VCF file and create some internal Python representation out of this (called `VariantTable`). Second, they take all reads from the given BAM file(s) and detect for each read which alleles it supports on which variant position. There are ways for this: If no reference is given, it is purely estimated from the CIGAR strings, which can be ambiguous in some cases, if multiple indels are too close together or even overlap. If a reference was given, a realignment is computed for each read (or parts of it) for variant it covers, where the read is a) aligned against the reference and b) against the reference but with one reference allele being replaced by the alternative from the VCF. In both cases, you get a `ReadSet` with `Read` objects, carrying the allele information.

After that, the actual phasing algorithm would be called. In your case you already have a partition of the reads (if I understood correctly) and would only compute a consensus sequence (0 or 1 for each variant position) for each partition. I don't know how the tag information is encoded into the BAM file, so getting the tag could be a bit more involved. If you want to create a PR for this feature, I would suggest to start with the file `whatshap/cli/phase.py`, because this is the starting point of the regular phasing command. The file is quite long and imports a lot of stuff, so the best would be to follow the main function and see what is done on a high-level to create a variant table and a read set.

A big question might be, where to put such a feature. One could either re-use the `phase`-command for this and add some flag like `--phase-from-read-tags` to tell the program to not run the phasing algorithm, but to phase from tags only. Or one creates a dedicated subcommand for this purpose, because it expects different input in some way. I would leave this for discussion with the @tobiasmarschall and @marcelm here.

Cheers,
Sven

**Comment by tobiasmarschall** [comment]
Thanks @schrins! In my view this is sufficiently different from `phase` to merit it's own subcommand. Almost everything needed should indeed be in the main phase function, which could be used as a template:
https://github.com/whatshap/whatshap/blob/cefdaececfbb1aa63176da301a0c13ff368aceb4/whatshap/cli/phase.py#L282
As an alternative to a tagged BAM file you could also read a simple tab-separated table assigning read names to a haplotype, similar to what is output by `whatshap haplotag --output-haplotag-list`.
@mrvollger Just ping us if anything is unclear.

**Comment by marcelm** [comment]
I agree this sounds more like a new subcommand, and I want to confirm that a PR adding this would be quite welcome.

See also our somewhat incomplete [development instructions](https://whatshap.readthedocs.io/en/latest/develop.html).

**Comment by nkkarpov** [comment]
Hi @mrvollger, I have opened the pull request. Currently, it works with basic functionality but lacks documentation. It implements a majority rule where votes are collected for phasing each variant based on the quality of the assigned alleles. I also tried to avoid assigning phase information inside long homopolymers. https://github.com/whatshap/whatshap/pull/521
